tests/end_to_end/workflow/reference_flow.py

```diff
@@ -0,0 +1,297 @@
+        self.collab_attr_dict_one = {key: key for key in range(5)}
+        self.collab_attr_file_one = io.StringIO("Test file data in collaborator")
+        self.collab_attr_math_one = math.sqrt(self.index)
+        self.collab_attr_complex_num_one = complex(self.index, self.index)
+        self.collab_attr_log_one = logging.getLogger(
+            "Test logger data in collaborator " + self.input
+        )
+
+        # append attributes of collaborator
+        TestFlowReference.step_one_collab_attrs.append(self)
+
+        if len(TestFlowReference.step_one_collab_attrs) >= 2:
+            collab_attr_list = filter_attrs(inspect.getmembers(self))
+            matched_ref_dict = find_matched_references(
+                collab_attr_list, TestFlowReference.step_one_collab_attrs
+            )
+            validate_collab_references(matched_ref_dict)
+
+        self.next(self.test_create_more_collab_attr)
+
+    @collaborator
+    def test_create_more_collab_attr(self):
+        """
+        Create different types of objects.
+        """
+
+        self.collab_attr_int_two = 30 + self.index
+        self.collab_attr_str_two = "String reference three " + self.input
+        self.collab_attr_list_two = [1, 2, 3, 5, 6, 8]
+        self.collab_attr_dict_two = {key: key for key in range(5)}
+        self.collab_attr_file_two = io.StringIO("Test file reference one")
+        self.collab_attr_math_two = math.sqrt(2)
+        self.collab_attr_complex_num_two = complex(2, 3)
+        self.collab_attr_log_two = logging.getLogger(
+            "Test logger data in collaborator" + self.input
+        )
+
+        TestFlowReference.step_two_collab_attrs.append(self)
+
+        if len(TestFlowReference.step_two_collab_attrs) >= 2:
+            collab_attr_list = filter_attrs(inspect.getmembers(self))
+            matched_ref_dict = find_matched_references(
+                collab_attr_list, TestFlowReference.step_two_collab_attrs
+            )
+            validate_collab_references(matched_ref_dict)
+
+        self.next(self.join)
+
+    @aggregator
+    def join(self, inputs):
+        """
+        Iterate over the references of collaborator attributes
+        validate uniqueness of attributes and raise assertion
+        """
+
+        all_attr_list = filter_attrs(inspect.getmembers(inputs[0]))
+        agg_attrs = filter_attrs(inspect.getmembers(self))
+
+        # validate aggregator references are intact after coming out of collaborators.
+        validate_agg_attr_ref(agg_attrs, self)
+
+        # validate collaborators references are not shared in between.
+        matched_ref_dict = find_matched_references(all_attr_list, inputs)
+        validate_collab_references(matched_ref_dict)
+
+        # validate aggregator references are not shared with any of the collaborators.
+        validate_agg_collab_references(inputs, self, agg_attrs)
+
+        all_shared_attr = ""
+        log.info("Reference test summary:")
```

Comment:
Is this logging needed?

Reply:
These logs are needed if we use Metaflow.